core/drivers/regulator/regulator_dt.c

```diff
@@ -0,0 +1,392 @@
+				 struct regulator *regulator)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	FMSG("Regulator: finalize %s registering", regulator_name(regulator));
+
+	res = regulator_register(regulator);
+	if (res)
+		return res;
+
+	if (regulator->ops->supplied_init) {
+		res = regulator->ops->supplied_init(regulator, fdt, node);
+		if (res)
+			return res;
+	}
+
+	return regulator_register_provider(fdt, node, regulator);
+}
+
+/*
+ * Pending regulators list: stores all regulator devices registered by their
+ * driver but not yet available to consumers as their dependency on their
+ * regulator supply is not yet resolved (supply has not been initialized yet).
+ */
+
+static void __maybe_unused print_pending_regulators(void)
+{
+	struct pending_regu *pending = NULL;
+
+	SLIST_FOREACH(pending, &pending_regu_list, link)
+		DMSG("Pending regulator %s",
+		     regulator_name(pending->regulator));
+}
+
+/*
+ * Returns true if at least 1 regulator found its supply and finalized its
+ * registration.
+ */
+static bool process_pending_list(void)
+{
+	struct dt_driver_provider *p = NULL;
+	struct pending_regu *pending = NULL;
+	struct pending_regu *next = NULL;
+	bool supplied = false;
+
+	SLIST_FOREACH_SAFE(pending, &pending_regu_list, link, next) {
+		p = dt_driver_get_provider_by_phandle(pending->supply_phandle,
+						      DT_DRIVER_REGULATOR);
+		if (!p)
+			continue;
+
+		pending->regulator->supply = dt_driver_provider_priv_data(p);
+
+		if (register_final(pending->fdt, pending->node,
+				   pending->regulator))
+			panic();
+
+		SLIST_REMOVE(&pending_regu_list, pending, pending_regu, link);
+		free(pending);
+
+		supplied = true;
+	}
+
+	return supplied;
+}
+
+/*
+ * Attempt to register pending regulators once their supply if found.
+ * Returns true if pending regulator list is empty upon processing.
+ */
+static bool resolve_pending_list(void)
+{
+	while (process_pending_list())
+		;
+
+	return SLIST_EMPTY(&pending_regu_list);
+}
+
+/* Adds a regulator to the pending list: those waiting for their supply */
+static TEE_Result add_to_pending_list(const void *fdt, int node,
+				      struct regulator *regulator,
+				      uint32_t supply_phandle,
+				      bool regulator_allocated)
+{
+	struct pending_regu *pending = NULL;
+
+	pending = calloc(1, sizeof(*pending));
+	if (!pending)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	*pending = (struct pending_regu){
+		.fdt = fdt,
+		.node = node,
+		.supply_phandle = supply_phandle,
+		.regulator = regulator,
+		.regulator_allocated = regulator_allocated,
+	};
+
+	SLIST_INSERT_HEAD(&pending_regu_list, pending, link);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result parse_dt(const void *fdt, int node,
+			   struct regulator *regulator)
+{
+	struct regulator_property *fp = NULL;
+	const fdt32_t *cuint = NULL;
+	int len = 0;
+
+	FMSG("Regulator: parse DT node %s", fdt_get_name(fdt, node, NULL));
+
+	regulator->name = fdt_getprop(fdt, node, "regulator-name", NULL);
```

Comment:
to fix, as reported by @GseoC. If "regulator-name" property name not found, preserved the name passed in `struct regu_dt_desc`.